File: javascript/0977-squares-of-a-sorted-array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortedSquares = function (nums) {
    let left = 0;
    let right = nums.length - 1;

    const answer = [];

    while (left <= right) {
        const leftSqr = Math.pow(nums[left], 2);
        const rightSqr = Math.pow(nums[right], 2);

        if (leftSqr > rightSqr) {
            answer.push(leftSqr);
            left++;
        } else {
            answer.push(rightSqr);
            right--;
        }
    }
    return answer.reverse();
};

File: kotlin/0977-squares-of-a-sorted-array.kt
class Solution {
    fun sortedSquares(nums: IntArray): IntArray {
        var left = 0; var right = nums.size-1; var end = nums.size-1
        val rArray = IntArray(nums.size)        
        while(left <= right){ // or while(end >= 0)
            val ls = nums[left] * nums[left]
            val rs = nums[right] * nums[right] 
            if(ls > rs){
                rArray[end] = ls
                left++
            }else{
                //rs > ls
                rArray[end] = rs
                right--
            }
            end--
        }
        return rArray
    }
}
